src/main/java/appeng/util/item/AESharedItemStack.java

```diff
@@ -0,0 +1,195 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Applied Energistics 2.  If not, see <http://www.gnu.org/licenses/lgpl>.
+ */
+
+package appeng.util.item;
+
+
+import java.util.Objects;
+
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+
+import appeng.api.config.FuzzyMode;
+
+
+public final class AESharedItemStack implements Comparable<AESharedItemStack>
+{
+	private static final NBTTagCompound LOW_TAG = new NBTTagCompound();
+	private static final NBTTagCompound HIGH_TAG = new NBTTagCompound();
+
+	private ItemStack itemStack;
+
+	public AESharedItemStack( final ItemStack itemStack )
+	{
+		this.itemStack = itemStack;
+	}
+
+	public ItemStack getItemStack()
+	{
+		return this.itemStack;
+	}
+
+	@Override
+	public int hashCode()
+	{
+		return Objects.hash( itemStack.getItem(), itemStack.getItemDamage() );
+	}
+
+	@Override
+	public boolean equals( final Object obj )
+	{
+		if( obj instanceof AESharedItemStack )
+		{
+			final AESharedItemStack other = (AESharedItemStack) obj;
+
+			if( this.itemStack == other.itemStack )
+			{
+				return true;
+			}
+			return ItemStack.areItemStacksEqual( this.itemStack, other.itemStack );
+		}
+		return false;
+	}
+
+	@Override
+	public int compareTo( AESharedItemStack b )
+	{
+		if( this.itemStack == b.itemStack )
+		{
+			return 0;
+		}
+
+		final int id = Item.getIdFromItem( this.itemStack.getItem() ) - Item.getIdFromItem( b.itemStack.getItem() );
+		if( id != 0 )
+		{
+			return id;
+		}
+
+		final int damageValue = this.itemStack.getItemDamage() - b.itemStack.getItemDamage();
+		if( damageValue != 0 )
+		{
+			return damageValue;
+		}
+
+		return this.compareNBT( b );
+	}
+
+	private int compareNBT( final AESharedItemStack b )
+	{
+		if( this.itemStack.getTagCompound() == b.itemStack.getTagCompound() )
+		{
+			return 0;
+		}
+
+		if( this.itemStack.getTagCompound() == LOW_TAG )
+		{
+			return -1;
+		}
+		if( this.itemStack.getTagCompound() == HIGH_TAG )
+		{
+			return 1;
+		}
+
+		final int nbt = ( this.itemStack == null ? 0 : this.itemStack.hashCode() ) - ( b.itemStack == null ? 0 : b.itemStack.hashCode() );
+		if( nbt != 0 )
+		{
+			return nbt;
+		}
+
+		return System.identityHashCode( this.itemStack.getTagCompound() ) - System.identityHashCode( b.itemStack.getTagCompound() );
+	}
+
+	AESharedItemStack getLow( final FuzzyMode fuzzy, final boolean ignoreMeta )
+	{
+		final ItemStack newDef = this.itemStack.copy();
+
+		if( ignoreMeta )
+		{
+			newDef.setItemDamage( 0 );
+		}
+		else
+		{
+			if( newDef.getItem().isDamageable() )
+			{
+				if( fuzzy == FuzzyMode.IGNORE_ALL )
+				{
+					newDef.setItemDamage( 0 );
+				}
+				else if( fuzzy == FuzzyMode.PERCENT_99 )
+				{
+					if( this.itemStack.getItemDamage() == 0 )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this conditional structure or edit its code blocks so that they're not all the same. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS3923)